### PR TITLE
fix(core): fixed fanout race condition

### PR DESCRIPTION
**Pull request by bluestreak01**

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 19 / 21 (90.48%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/FanOut.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 19 (89.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/FanOut.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 19 (89.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/FanOut.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|

File: core/src/main/java/io/questdb/mp/RingQueue.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
 
     @SuppressWarnings("unchecked")
     public RingQueue(ObjectFactory<T> factory, int cycle) {
+        assert cycle < 3 || cycle == (cycle / 2) * 2;
```
Comment (ideoma) [inline core/src/main/java/io/questdb/mp/RingQueue.java:38]: I don't think this is correct to check that cycle is power of 2. Should be something like `Integer.bitCount(cycle) < 2`

Comment (eugenels) [reply core/src/main/java/io/questdb/mp/RingQueue.java:38]: (cycle & (cycle - 1) == 0, if cycle != 0

Comment (bluestreak01) [reply core/src/main/java/io/questdb/mp/RingQueue.java:38]: Thank you champs 👍

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/FanOut.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/FanOut.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 69 (89.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/OpenBarrier.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/mp/FanOut.java|14|20|70.00%|
|:large_blue_circle:|io/questdb/mp/SPSequence.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/MessageBusImpl.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractSSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractMSequence.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/network/AbstractIODispatcher.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 69 (89.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/OpenBarrier.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/mp/FanOut.java|14|20|70.00%|
|:large_blue_circle:|io/questdb/mp/SPSequence.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/MessageBusImpl.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractSSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractMSequence.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/network/AbstractIODispatcher.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 69 (89.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/OpenBarrier.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/mp/FanOut.java|14|20|70.00%|
|:large_blue_circle:|io/questdb/mp/SPSequence.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/MessageBusImpl.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractSSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractMSequence.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/network/AbstractIODispatcher.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 69 (89.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/OpenBarrier.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/mp/FanOut.java|14|20|70.00%|
|:large_blue_circle:|io/questdb/mp/SPSequence.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/MessageBusImpl.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractSSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractMSequence.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/network/AbstractIODispatcher.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 69 (89.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/mp/OpenBarrier.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/mp/FanOut.java|14|20|70.00%|
|:large_blue_circle:|io/questdb/mp/SPSequence.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/MessageBusImpl.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractSSequence.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SCSequence.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/mp/RingQueue.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/mp/AbstractMSequence.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/network/AbstractIODispatcher.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|1|1|100.00%|

**Review by ideoma — APPROVED** [review]

**Pull request merged**